### PR TITLE
Defined application names for all database sessions. Closes #3600

### LINKED ISSUE #3600: Specific `application_name` for Connections to the service by the system
There are a few queries that need to be sent to the service solely for system usage. Examples are:
* load `connection_state`
* load `server_settings`
etc.

Connections which are used to send these queries should have a distinct application name so that it's easy to discern these queries from user initiated queries

### PR DESCRIPTION
**Pull request by binaek**

### CODE REVIEW

File: pkg/db/db_client/db_client.go [hunk 1]
```diff
@@ -109,13 +111,14 @@ func NewDbClient(ctx context.Context, connectionString string, onConnectionCallb
 	return client, nil
 }
 
-func (c *DbClient) loadServerSettings(ctx context.Context) error {
-	conn, _, err := c.GetDatabaseConnectionWithRetries(ctx)
-	if err != nil {
-		return err
+func (c *DbClient) closePools(ctx context.Context) {
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client.go:114]: ctx not needed

rename to `closePool`
would pool ever be nil?

Comment (kaidaguerre) [reply pkg/db/db_client/db_client.go:114]: dbclient creation fails if pool creation fails, so we can never have a db client without a pool, right? 

So null check is not necessary?

File: pkg/db/db_client/db_client_connect.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ func (c *DbClient) establishConnectionPool(ctx context.Context) error {
 		"localhost",
 	}
 
+	// this will yield a false negative when connecting to a local service using a network IP
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client_connect.go:40]: and...? Is this bad? do we need to handle it?

File: pkg/db/db_client/db_client_connect.go [hunk 3]
```diff
@@ -55,8 +56,10 @@ func (c *DbClient) establishConnectionPool(ctx context.Context) error {
 	// set an app name so that we can track database connections from this Steampipe execution
 	// this is used to determine whether the database can safely be closed
 	config.ConnConfig.Config.RuntimeParams = map[string]string{
-		"application_name": runtime.PgClientAppName,
+		constants.RuntimeParamsKeyApplicationName: runtime.ClientConnectionAppName,
 	}
+	// disable automatic prepared statements
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client_connect.go:61]: is there a performance cost associated with disabling this?

File: pkg/db/db_client/db_client_execute.go [hunk 4]
```diff
@@ -199,53 +200,45 @@ func (c *DbClient) getQueryTiming(ctx context.Context, startTime time.Time, sess
...
-		id = rw.Data[0].(int64)
-		rowsFetched := rw.Data[1].(int64)
-		cacheHit := rw.Data[2].(bool)
-		hydrateCalls := rw.Data[3].(int64)
-
-		timingResult.Metadata.HydrateCalls += hydrateCalls
-		if cacheHit {
-			timingResult.Metadata.CachedRowsFetched += rowsFetched
+	timingResult.Metadata = &queryresult.TimingMetadata{}
+	for _, smr := range scanRows {
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client_execute.go:223]: `r` is better than `smr`

File: pkg/db/db_client/db_client_execute.go [hunk 5]
```diff
@@ -199,53 +200,45 @@ func (c *DbClient) getQueryTiming(ctx context.Context, startTime time.Time, sess
 		resultChannel <- timingResult
 	}()
 
-	res, err := c.ExecuteSyncInSession(ctx, session, fmt.Sprintf("select id, rows_fetched, cache_hit, hydrate_calls from %s.%s where id > %d", constants.InternalSchema, constants.ForeignTableScanMetadata, session.ScanMetadataMaxId))
+	var scanRows []ScanMetadataRow
+	_ = db_common.ExecuteSystemClientCallOnConnection(ctx, session.Connection.Conn(), func(ctx context.Context, tx pgx.Tx) error {
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client_execute.go:204]: comment

Comment (kaidaguerre) [reply pkg/db/db_client/db_client_execute.go:204]: put query into a variable

File: pkg/db/db_client/db_client_execute.go [hunk 6]
```diff
@@ -199,53 +200,45 @@ func (c *DbClient) getQueryTiming(ctx context.Context, startTime time.Time, sess
...
+		rows, err := tx.Query(ctx, fmt.Sprintf("select id, rows_fetched, cache_hit, hydrate_calls from %s.%s where id > %d", constants.InternalSchema, constants.ForeignTableScanMetadata, session.ScanMetadataMaxId))
+		if err != nil {
+			return err
+		}
+		collected, err := pgx.CollectRows(rows, pgx.RowToStructByName[ScanMetadataRow])
+		scanRows = collected
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client_execute.go:210]: can't we read directly into scanRows?

File: pkg/db/db_common/cache_control.go [hunk 7]
```diff
@@ -32,21 +32,25 @@ func SetCacheEnabled(ctx context.Context, enabled bool, connection *pgx.Conn) er
...
 func executeCacheSetFunction(ctx context.Context, settingValue string, connection *pgx.Conn) error {
-	_, err := connection.Exec(ctx, fmt.Sprintf(
-		"select %s.%s('%s')",
-		constants.InternalSchema,
-		constants.FunctionCacheSet,
-		settingValue,
-	))
-	return err
+	return ExecuteSystemClientCallOnConnection(ctx, connection, func(ctx context.Context, tx pgx.Tx) error {
+		_, err := connection.Exec(ctx, fmt.Sprintf(
```
Comment (kaidaguerre) [inline pkg/db/db_common/cache_control.go:36]: why do we sometimes use the connection and sometimes the tx?

File: pkg/db/db_common/session_system.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	"github.com/turbot/steampipe/pkg/constants"
+	"github.com/turbot/steampipe/pkg/constants/runtime"
+)
+
+// SystemClientExecutor is the executor function that is called within a transaction
+// make sure that by the time the executor finishes execution, the connection is freed
+// otherwise we will get a `conn is busy` error
+type SystemClientExecutor func(context.Context, pgx.Tx) error
+
+// ExecuteSystemClientCallOnConnection creates a transaction and sets the application_name to the
```
Comment (kaidaguerre) [inline pkg/db/db_common/session_system.go:17]: `ExecuteSystemClientCallOnConnection` -> `ExecuteSystemClientCall`

File: pkg/db/db_common/session_system.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+// one used by the system client, executes the callback and sets the application name back to the client app name
+func ExecuteSystemClientCallOnConnection(ctx context.Context, conn *pgx.Conn, executor SystemClientExecutor) error {
+	// checks that the appname is the one reserved for user-originating queries
+	appNameNeedsUpdate := IsClientAppName(conn.Config().RuntimeParams[constants.RuntimeParamsKeyApplicationName])
```
Comment (kaidaguerre) [inline pkg/db/db_common/session_system.go:21]: why is this needed?

Comment (kaidaguerre) [reply pkg/db/db_common/session_system.go:21]: surely we can assume the appname will always be the client app name?

File: pkg/db/db_local/execute.go [hunk 10]
```diff
@@ -50,26 +44,17 @@ func ExecuteSqlInTransaction(ctx context.Context, conn *pgx.Conn, statements ...
...
 
 func ExecuteSqlWithArgsInTransaction(ctx context.Context, conn *pgx.Conn, queries ...db_common.QueryWithArgs) (results []pgconn.CommandTag, err error) {
-	tx, err := conn.Begin(ctx)
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		if err != nil {
-			tx.Rollback(ctx)
+	err = pgx.BeginFunc(ctx, conn, func(tx pgx.Tx) error {
```
Comment (kaidaguerre) [inline pkg/db/db_local/execute.go:47]: why update this but not `ExecuteSqlInTransaction`?

File: pkg/connection/refresh_connections_state.go [hunk 11]
```diff
@@ -53,7 +54,7 @@ type refreshConnectionState struct {
 func newRefreshConnectionState(ctx context.Context, forceUpdateConnectionNames []string) (*refreshConnectionState, error) {
 	// create a connection pool to connection refresh
```
Comment (kaidaguerre) [inline pkg/connection/refresh_connections_state.go:55]: comment  on appname

File: pkg/connection/refresh_connections_state.go [hunk 12]
```diff
@@ -754,7 +755,7 @@ func (s *refreshConnectionState) setIncompleteConnectionStateToError(ctx context
 
 // OnConnectionsChanged is the callback function invoked by the connection watcher when connections are added or removed
 func (s *refreshConnectionState) sendPostgreSchemaNotification(ctx context.Context) error {
-	conn, err := db_local.CreateLocalDbConnection(ctx, &db_local.CreateDbOptions{Username: constants.DatabaseSuperUser})
+	conn, err := db_local.CreateLocalDbConnection(ctx, &db_local.CreateDbOptions{Username: constants.DatabaseSuperUser, AppName: runtime.ServiceConnectionAppName})
```
Comment (kaidaguerre) [inline pkg/connection/refresh_connections_state.go:758]: comment on appname

Comment (kaidaguerre) [reply pkg/connection/refresh_connections_state.go:758]: why a separate connection

File: pkg/db/db_client/db_client.go [hunk 13]
```diff
@@ -157,13 +164,12 @@ func (c *DbClient) ServerSettings() *db_common.ServerSettings {
 
 // Close implements Client
 // closes the connection to the database and shuts down the backend
-func (c *DbClient) Close(context.Context) error {
+func (c *DbClient) Close(ctx context.Context) error {
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client.go:167]: remove ctx

File: pkg/db/db_client/db_client.go [hunk 14]
```diff
@@ -188,7 +194,7 @@ func (c *DbClient) RefreshSessions(ctx context.Context) (res *db_common.AcquireS
 // NOTE: it optimises the schema extraction by extracting schema information for
 // connections backed by distinct plugins and then fanning back out.
 func (c *DbClient) GetSchemaFromDB(ctx context.Context) (*db_common.SchemaMetadata, error) {
-	conn, _, err := c.GetDatabaseConnectionWithRetries(ctx)
+	conn, err := c.sysPool.Acquire(ctx)
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client.go]: comment - get connection for management pool

File: pkg/db/db_client/db_client_connect.go [hunk 15]
```diff
@@ -74,5 +80,29 @@ func (c *DbClient) establishConnectionPool(ctx context.Context) error {
 		return err
 	}
 	c.pool = dbPool
+
+	return c.establishSystemConnectionPool(ctx, config)
+}
+
+// establishSystemConnectionPool creates a connection pool that can be used to support execution of user-initiated
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client_connect.go:87]: // establishSystemConnectionPool creates a connection pool to use to execute 
// system-initiated queries (loading of connection state etc.)
// unlike establishConnectionPool, which is run first to create the user-query pool
// this doesn't wait for the pool to completely start, as establishConnectionPool will have established and verified a connection with the service
func (c *DbClient) establishSystemConnectionPool(ctx context.Context, config *pgxpool.Config) error {

File: pkg/db/db_common/session_system.go [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+// SystemClientExecutor is the executor function that is called within a transaction
+// make sure that by the time the executor finishes execution, the connection is freed
+// otherwise we will get a `conn is busy` error
+type SystemClientExecutor func(context.Context, pgx.Tx) error
+
+// ExecuteSystemClientCall creates a transaction and sets the application_name to the
+// one used by the system client, executes the callback and sets the application name back to the client app name
+func ExecuteSystemClientCall(ctx context.Context, conn *pgx.Conn, executor SystemClientExecutor) error {
+	if !IsClientAppName(conn.Config().RuntimeParams[constants.RuntimeParamsKeyApplicationName]) {
+		// this should NEVER happen
```
Comment (kaidaguerre) [inline pkg/db/db_common/session_system.go:22]: return error

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: pkg/interactive/interactive_client.go [hunk 17]
```diff
@@ -723,7 +722,7 @@ func (c *InteractiveClient) listenToPgNotifications(ctx context.Context) error {
 		}
 
 		log.Printf("[TRACE] Wait for notification")
-		notification, err := conn.WaitForNotification(ctx)
+		notification, err := conn.Conn().WaitForNotification(ctx)
```
Comment (kaidaguerre) [inline pkg/interactive/interactive_client.go:725]: this notification connection should be system pool I think

File: pkg/connection/refresh_connections_state.go [hunk 18]
```diff
@@ -754,12 +749,11 @@ func (s *refreshConnectionState) setIncompleteConnectionStateToError(ctx context
 
 // OnConnectionsChanged is the callback function invoked by the connection watcher when connections are added or removed
 func (s *refreshConnectionState) sendPostgreSchemaNotification(ctx context.Context) error {
-	conn, err := db_local.CreateLocalDbConnection(ctx, &db_local.CreateDbOptions{Username: constants.DatabaseSuperUser})
+	conn, err := s.pool.Acquire(ctx)
```
Comment (kaidaguerre) [inline pkg/connection/refresh_connections_state.go:752]: is this a system pool
we can use the system pool for notifications

File: pkg/interactive/interactive_client.go [hunk 19]
```diff
@@ -735,14 +734,14 @@ func (c *InteractiveClient) listenToPgNotifications(ctx context.Context) error {
...
+	conn.Release()
 
 	log.Printf("[TRACE] InteractiveClient listenToPgNotifications DONE")
 	return nil
 }
 
-func (c *InteractiveClient) getNotificationConnection(ctx context.Context) (*pgx.Conn, error) {
-	conn, err := db_local.CreateLocalDbConnection(ctx, &db_local.CreateDbOptions{Username: constants.DatabaseUser})
+func (c *InteractiveClient) getNotificationConnection(ctx context.Context) (*pgxpool.Conn, error) {
+	conn, err := c.client().AcquireConnection(ctx)
```
Comment (kaidaguerre) [inline pkg/interactive/interactive_client.go:744]: use system pool

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: pkg/db/db_client/db_client.go (hunk 13 above)
Comment (binaek) [reply pkg/db/db_client/db_client.go:167]: We can't remove this context since the local implementation of DBClient uses the context for service shutdown

File: pkg/connection/refresh_connections_state.go [hunk 20]
```diff
@@ -33,7 +33,8 @@ type connectionError struct {
 }
 
 type refreshConnectionState struct {
-	pool                       *pgxpool.Pool
+	// a connection pool to the DB service which uses the server appname
+	refreshConnectionPool      *pgxpool.Pool
```
Comment (kaidaguerre) [inline pkg/connection/refresh_connections_state.go:37]: just call it `pool`

Comment (kaidaguerre) [reply pkg/connection/refresh_connections_state.go:37]: we know that all queries refreshConnectionState does will use an admin pool

File: pkg/db/db_client/db_client.go (hunk 13 above)
Comment (kaidaguerre) [reply pkg/db/db_client/db_client.go:168]: remove the unused var name then

File: pkg/db/db_client/db_client_session.go [hunk 21]
```diff
@@ -13,23 +12,18 @@ import (
 )
 
 func (c *DbClient) AcquireConnection(ctx context.Context) (*pgxpool.Conn, error) {
```
Comment (kaidaguerre) [inline pkg/db/db_client/db_client_session.go:14]: who uses these functions?

File: pkg/db/db_common/client.go [hunk 22]
```diff
@@ -17,6 +17,8 @@ type Client interface {
 
 	// acquire a database connection - must be closed
 	AcquireConnection(ctx context.Context) (*pgxpool.Conn, error)
```
Comment (kaidaguerre) [inline pkg/db/db_common/client.go:19]: remove all var names

File: pkg/db/db_local/install.go [hunk 23]
```diff
@@ -317,15 +317,8 @@ func runInstall(ctx context.Context, oldDbName *string) error {
...
-	firstCharacter := databaseName[0:1]
-	var ascii int
-	for _, r := range databaseName {
-		ascii = int(r)
-		break
-	}
-	if firstCharacter == "_" || (ascii >= 'a' && ascii <= 'z') {
-		log.Printf("[TRACE] valid database name: %s", databaseName)
-	} else {
+	validDBName := databaseName[0] == '-' || (databaseName[0] >= 'a' && databaseName[0] <= 'z')
```
Comment (kaidaguerre) [inline pkg/db/db_local/install.go:320]: prev logic checked for `_`  as first char, you've changed it to `-`

Comment (kaidaguerre) [reply pkg/db/db_local/install.go:320]: add function IsValidDbName with tests

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

**Review by kaidaguerre — APPROVED** [review]
